### ISSUE TITLE
Out of order compound dtypes

### ISSUE THREAD

**Issue by ngoldbaum** [h5py/h5py#782]
It looks the way h5py is using numpy's dtype constuctor is causing issues with hdf5 files that have out-of-order dtypes. This is causing issues. See #764, #765, #738, as well as @nevion's comment: https://github.com/h5py/h5py/pull/765#issuecomment-263463634. It's possible this is a numpy bug rather than an h5py bug, but h5py will still need to work around it given the existence of old versions of numpy in the wild.

**Comment by aragilar** [comment]
This should be fixed in the latest master by #793.

**Comment by nevion** [comment]
@aragilar bump for re-eval - for #765 specifically - #793 focuses on generating a correct h5py compound type for a given numpy dtype, but #765 focused on generating a correct dtype for a given h5py - they're complementary.  I just want to make sure, as it seemed that we caught numpy's dtype constructor barfling on out of order fields also, that we have both cases covered and tested working.

**Comment by aragilar** [comment]
@nevion I've created #802 which is an implementation of the test you suggested (I thought that was already added in a previous request, but apparently not), is there anything else you think isn't covered by #793 or #802?

**Comment by nevion** [comment]
one more thing before you close this out, in #802, can you make it an unpacked struct or test one of those in addition?

**Comment by aragilar** [comment]
@nevion I'm confused how you can have an unpacked struct (which I understand as a struct whose elements are aligned based on CPU arch/type size, as opposed to packed where the elements are next to each other without padding, I fully accept this may be superficial and/or wrong) which is out-of-order, given that to have the struct be out-of-order you must define offsets, which seems the antithesis of unpacked struct. Maybe an example would clear up my confusion? In any case, I'm going to reopen this until we've finished and merged #802.

**Comment by nevion** [comment]
Consider a couple cases (and the potential for real world mixing of them):

- a user who reorders data fields inside a struct, after writing compound type wrapper code.  That is the offsets change later on.  The API does not prohibit this and it's not a bad thing at all to preserve the order of how things were appended to the compound type's field list, but it might trip some people up, who expected monotonic offsets.  This case would frequently happen in refactors or insertion of new fields.
- a user who exports only a subset of struct fields
```
struct SensedPacket{
  int acc, usr1, usr2, diag, temp, val;
}
```
And the user only wants to export acc, and diag - for instance (forgoing compound type creation)


You may save a packed or unpacked struct directly to a dataset without the additional conversion pass (which would uses the type conversion engine, a fairly heavy weight copy) - this is what the packet tables api allows (which is just sitting on top of H5Tables' implementation of append, thus is technically there too).

**Comment by tacaswell** [comment]
What is the state of this?

**Comment by takluyver** [comment]
@aragilar reopened this until #802 was merged, which it was before 2.7. I think @nevion was asking for an additional test, but it's not obvious to me how to write that. I suggest that we close this issue again,  and ask @nevion to make a new issue (or better, PR ;-) if the additional test is still relevant.

**Comment by nevion** [comment]
I'll look at adding a test, I don't understand what was difficult or confusing about exporting a subset of fields or creating the compound type with the list of fields not in the order of the struct described, but with the same offsets.

**Comment by tacaswell** [comment]
punted to 2.10

**Comment by takluyver** [comment]
Dropping the milestone because it seems like this isn't a priority for anyone. We can still do it for 2.10 if someone wants to get to it soon.

**Comment by nevion** [comment]
@takluyver please see https://github.com/h5py/h5py/pull/1214

**Comment by takluyver** [comment]
Closed by #1214.
